pynbody/halo/adaptahop.py

```diff
@@ -136,10 +142,35 @@ def _read_halo_data(self, halo_id, offset):
         halo : Halo object
             The halo object, filled with the data read from file.
         """
+
+        default_dtype = getattr(self.base, "_iord_dtype", "i")
+        possible_dtypes = list({"i", "q"} - {default_dtype})
+
+        dtypes = [default_dtype] + possible_dtypes
+
+        def _helper(fpu, expected_size):
+            ipos = fpu.tell()
+
+            for dtype in dtypes:
+                try:
+                    iord_array = fpu.read_vector(dtype)
```

Comment:
Is this a bit strange - repeatedly reading the file and allowing it to fail? Would it be easy/more streamlined to actually look at the fortran reported block size and infer the dtype from that directly?

Reply:
This code should only fail on the first read. After this, the right dtype is stored (so there should not be any significant performance hit, and worst case is that the data will be read twice).

It is possible in principle to read the header of the Fortran record and infer the dtype given the expected number of elements to be read, but it requires having a fairly low-level piece of code there. Another option is to add support for this directly in the Fortran file reader (in `fpu.read_vector`) by adding the possibility to specify a length and possible data types.

Reply:
OK, fair enough! Thanks